### ISSUE TITLE
Compatibility with OpenTelemetry - Missing "verbose" method

### ISSUE THREAD

**Issue by NPellet** [log4js-node/log4js-node#1339]
I would very much like to use log4js as a logger for OpenTelemetry in the following way:

```typescript
import opentelemetry, { DiagConsoleLogger, DiagLogLevel } from '@opentelemetry/api'
import log4js from 'log4js'
const serviceLogger = log4js.getLogger('logger')

opentelemetry.diag.setLogger(serviceLogger, DiagLogLevel.DEBUG) // <== Failing here
```

However, the logger interface is missing the `  verbose(message: any, ...args: any[]): void;` class member signature.

I believe that with this method, log4js would be compatible with otel (at least the current beta specification)

I'd be happy to write a PR if that'd be acceptable

**Comment by lamweili** [comment]
Hi, there are a few more signature differences (apart from the levels) between `log4js` and `OpenTelemetry`:
- `log4js` API: https://log4js-node.github.io/log4js-node/api.html
- `OpenTelemetry` API: https://open-telemetry.github.io/opentelemetry-js-api/classes/diagapi.html

Nevertheless, for log levels, there are a couple of ways to bridge between `log4js` and `OpenTelemetry`.

1. By mapping `logger.verbose` to `logger.trace`
(referencing [**log4js API**: Loggers - log4js.getLogger([category])](https://log4js-node.github.io/log4js-node/api.html)
```js
const log4js = require('log4js');
const logger = log4js.getLogger();

// sets default level (alternatively, provide `categories.default.level` during `log4js.configure()`)
logger.level = 'trace'; 

// maps verbose to trace
logger.verbose = logger.trace;
logger.verbose('hello world');
```
3. By creating a custom log level (specifically for OpenTelemetry)
(referencing [**log4js API**: configuration - log4js.configure(object || string)](https://log4js-node.github.io/log4js-node/api.html) and [levels.js#L95-L105](https://github.com/log4js-node/log4js-node/blob/b2c97040fb35b1a5ef64ea50b5a8c0d0c09886ab/lib/levels.js#L95-L105))

```js
const log4js = require('log4js');
log4js.configure({
  levels: {
    verbose: { value: 5000, colour: 'blue' },
  },
  appenders: {
    stdout: { type: 'stdout' },
  },
  categories: {
    default: { appenders: ['stdout'], level: 'verbose' },
  },
});

const logger = log4js.getLogger();
logger.verbose('hello world');
```

**Comment by lamweili** [comment]
Hi, there are a few more signature differences (apart from the levels) between `log4js` and `OpenTelemetry`:
- `log4js` API: https://log4js-node.github.io/log4js-node/api.html
- `OpenTelemetry` API: https://open-telemetry.github.io/opentelemetry-js-api/classes/diagapi.html

Nevertheless, for log levels, there are a couple of ways to bridge between `log4js` and `OpenTelemetry`.

1. By mapping `logger.verbose` to `logger.trace`
(referencing [**log4js API**: Loggers - log4js.getLogger([category])](https://log4js-node.github.io/log4js-node/api.html)
```js
const log4js = require('log4js');
const logger = log4js.getLogger();

// sets default level (alternatively, provide `categories.default.level` during `log4js.configure()`)
logger.level = 'trace'; 

// maps verbose to trace
logger.verbose = logger.trace;
logger.verbose('hello world');
```
2. By creating a custom log level (specifically for OpenTelemetry)
(referencing [**log4js API**: configuration - log4js.configure(object || string)](https://log4js-node.github.io/log4js-node/api.html) and [levels.js#L95-L105](https://github.com/log4js-node/log4js-node/blob/b2c97040fb35b1a5ef64ea50b5a8c0d0c09886ab/lib/levels.js#L95-L105))

```js
const log4js = require('log4js');
log4js.configure({
  levels: {
    verbose: { value: 5000, colour: 'blue' },
  },
  appenders: {
    stdout: { type: 'stdout' },
  },
  categories: {
    default: { appenders: ['stdout'], level: 'verbose' },
  },
});

const logger = log4js.getLogger();
logger.verbose('hello world');
```

**Comment by lamweili** [comment]
Hi, there are a few more signature differences (apart from the levels) between `log4js` and `OpenTelemetry`:
- `log4js` API: https://log4js-node.github.io/log4js-node/api.html
- `OpenTelemetry` API: https://open-telemetry.github.io/opentelemetry-js-api/classes/diagapi.html

Nevertheless, for log levels, there are a couple of ways to bridge between `log4js` and `OpenTelemetry`.

1. By mapping `logger.verbose` to `logger.trace`
(referencing [**log4js API**: Loggers - log4js.getLogger([category])](https://log4js-node.github.io/log4js-node/api.html)
```js
const log4js = require('log4js');
const logger = log4js.getLogger();

// sets default level (alternatively, provide `categories.default.level` during `log4js.configure()`)
logger.level = 'trace'; 

// maps verbose to trace
logger.verbose = logger.trace;
logger.verbose('hello world');
```
2. By creating a custom log level (specifically for OpenTelemetry)
(referencing [**log4js API**: configuration - log4js.configure(object || string)](https://log4js-node.github.io/log4js-node/api.html) and [levels.js#L95-L105](https://github.com/log4js-node/log4js-node/blob/b2c97040fb35b1a5ef64ea50b5a8c0d0c09886ab/lib/levels.js#L95-L105))

```js
const log4js = require('log4js');
log4js.configure({
  levels: {
    verbose: { value: 5000, colour: 'blue' }, // mirrors trace
  },
  appenders: {
    stdout: { type: 'stdout' },
  },
  categories: {
    default: { appenders: ['stdout'], level: 'verbose' },
  },
});

const logger = log4js.getLogger();
logger.verbose('hello world');
```

**Comment by NPellet** [comment]
I think the relevant interface is this one: https://open-telemetry.github.io/opentelemetry-js-api/interfaces/diaglogger.html so there would really be just this one method.

Your approaches are sound at a javascript level, but sadly neither or those approach pass static checking with typescript.

**Comment by lamweili** [comment]
I think approach 2 would suit your needs better in that case - to create a custom log level that is `verbose`.

**Comment by lamweili** [comment]
Would that work? Or what do you have in mind?

**Comment by NPellet** [comment]
Well it's like I said, I would break the types and that would violate our coding guidelines :(

But anyways, the open telemetry API is still evolving, I'd suggest waiting to see where it goes and close this for now

**Comment by lamweili** [comment]
The incorrect typing for custom levels is now mentioned in:
- #1347

**Comment by lamweili** [comment]
@NPellet, can you try out the solutions in https://github.com/log4js-node/log4js-node/issues/1347#issuecomment-1316722109?
The custom levels' typing issue should be resolved, and the type should no longer break.

**Comment by lamweili** [comment]
@NPellet, can you try out the solutions in https://github.com/log4js-node/log4js-node/issues/1347#issuecomment-1316722109?
The custom levels' typing issue should be resolved, and the type should no longer break.

**Comment by lamweili** [comment]
Released in `log4js@6.7.1`.

**Comment by lamweili** [comment]
~~https://open-telemetry.github.io/opentelemetry-js-api/interfaces/diaglogger.html~~ is now 
 https://open-telemetry.github.io/opentelemetry-js/interfaces/_opentelemetry_api.DiagLogger.html
